### PR TITLE
Fix mines

### PR DESCRIPTION
**Pull request by StrawWagen**
Fixed landmines bypassing CPPI
Moved claymore det key off of keypad.

### CODE REVIEW

**Comment by StrawWagen** [comment]
Not super happy with my implementation of JID.CanConstrain But it's all i got.

**Comment by wrefgtzweve** [comment]
You shouldnt constantly have to check if it's world AND if it's allowed, the world check should be moved within JID.CanConstrain

File: lua/autorun/server/ji_dsserver.lua [hunk 1]
```diff
@@ -105,6 +105,22 @@ function JID.DetermineAttacker( ent )
     return ent
 end
 
+function JID.CanConstrain( checker, toConstrain )
```
Comment (wrefgtzweve) [inline lua/autorun/server/ji_dsserver.lua]: rename "checker" here to "ent", checker sounds like a function of some sort

File: lua/autorun/server/ji_dsserver.lua [hunk 2]
```diff
@@ -105,6 +105,22 @@ function JID.DetermineAttacker( ent )
...
+    if not CPPI then return true end
+
+    if not IsValid( checker ) then return end
+
+    local checkerOwner = checker:CPPIGetOwner()
+    local toConstrainOwner = toConstrain:CPPIGetOwner()
+    local isWorld = toConstrain:IsWorld()
```
Comment (wrefgtzweve) [inline lua/autorun/server/ji_dsserver.lua]: You're checking for world twice, in this function and in the entities themselves

**Comment by StrawWagen** [comment]
Each sent reacting differently if it's trying to interact with the world vs a prop, is a behavior i want to carry over, ill do some thinking for the function.

File: lua/autorun/server/ji_dsserver.lua [hunk 3]
```diff
@@ -105,6 +105,19 @@ function JID.DetermineAttacker( ent )
...
+
+    if not IsValid( ent ) then return end
+    if not IsValid( toConstrain ) then return end
+
+    local toConstrainOwner = toConstrain:CPPIGetOwner()
+    if not IsValid( toConstrainOwner ) then return end
+
+    local entOwner = ent:CPPIGetOwner()
+    if entOwner == toConstrainOwner then return true end
+end
```
Comment (legokidlogan) [inline lua/autorun/server/ji_dsserver.lua]: Proper usage of CPPI; check edge cases
```suggestion
function JID.CanConstrain( ent, toConstrain )
    if not IsValid( ent ) then return end
    if not IsValid( toConstrain ) then return end

    local entOwner = CPPI and ent:CPPIGetOwner() or ent:GetOwner()
    if not IsValid( entOwner ) then return end
    if CPPI and not toConstrain:CPPICanTool( entOwner ) then return end

    local toConstrainOwner = CPPI and toConstrain:CPPIGetOwner() or toConstrain:GetOwner()
    if not IsValid( toConstrainOwner ) then return end
    if CPPI and not ent:CPPICanTool( toConstrainOwner ) then return end

    if CPPI then return true end -- CPPI exists and hasn't blocked, so allow
    if entOwner == toConstrainOwner then return true end -- CPPI doesn't exist, directly compare owners
end
```

**Review by legokidlogan — CHANGES_REQUESTED** [review]

File: lua/autorun/server/ji_dsserver.lua (hunk 3 above)
Comment (wrefgtzweve) [reply lua/autorun/server/ji_dsserver.lua]: can you add
```lua
local CPPI = CPPI 
```

Comment (StrawWagen) [reply lua/autorun/server/ji_dsserver.lua]: Done

File: lua/entities/ent_jack_claymore/init.lua [hunk 4]
```diff
@@ -145,7 +145,10 @@ function ENT:Use( activator )
         local Tr = util.QuickTrace( activator:GetShootPos(), activator:GetAimVector() * 100, { self, activator } )
 
         if Tr.Hit and IsValid( Tr.Entity:GetPhysicsObject() ) then
-            if table.HasValue( PlantableMats, Tr.MatType ) then
+            local canConstrain = JID.CanConstrain( self, Tr.Entity ) or Tr.Entity:IsWorld()
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_claymore/init.lua:156]: why would IsWorld be checked here when others dont?

File: lua/entities/ent_jack_claymore/init.lua [hunk 5]
```diff
@@ -145,7 +145,10 @@ function ENT:Use( activator )
         local Tr = util.QuickTrace( activator:GetShootPos(), activator:GetAimVector() * 100, { self, activator } )
 
         if Tr.Hit and IsValid( Tr.Entity:GetPhysicsObject() ) then
-            if table.HasValue( PlantableMats, Tr.MatType ) then
+            local canConstrain = JID.CanConstrain( self, Tr.Entity ) or Tr.Entity:IsWorld()
+
+            -- stick into loose mats solidly
+            if table.HasValue( PlantableMats, Tr.MatType ) and canConstrain then
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_claymore/init.lua]: don't use `table.HasValue` use a kv lookup table please

File: lua/entities/ent_jack_landmine/init.lua [hunk 6]
```diff
@@ -142,11 +142,13 @@ function ENT:Use( activator )
     self:SetAngles( traceResult.HitNormal:Angle() )
     self:SetPos( traceResult.HitPos + traceResult.HitNormal )
 
-    if traceResult.Entity == game.GetWorld() then
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_landmine/init.lua:145]: why was the world check removed

**Review by wrefgtzweve — CHANGES_REQUESTED** [review]

File: lua/entities/ent_jack_claymore/init.lua (hunk 4 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_claymore/init.lua:156]: claymore welds itself to the world if it's planted ontop of grass, other mines just freeze themself, i'd like to keep this functionality

File: lua/entities/ent_jack_landmine/init.lua (hunk 6 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_landmine/init.lua:145]: used to freeze if the ent was the world, else parent to it
now it parents if it can, else freezes

**Comment by StrawWagen** [comment]
Tested code, permissions work.

File: lua/entities/ent_jack_claymore/init.lua [hunk 7]
```diff
@@ -149,7 +156,7 @@ function ENT:Use( activator )
             local canConstrain = JID.CanConstrain( self, Tr.Entity ) or Tr.Entity:IsWorld()
 
             -- stick into loose mats solidly
-            local isPlantableMat = PlantableMats[Tr.MatType]
+            local isPlantableMat = plantableMats[Tr.MatType]
 
             if isPlantableMat and canConstrain then
```
Comment (wrefgtzweve) [inline lua/entities/ent_jack_claymore/init.lua]: no reason to have this as a variable, you can just directly put this in line 161

**Review by legokidlogan — APPROVED** [review]

**Review by wrefgtzweve — APPROVED** [review]

**Pull request merged**